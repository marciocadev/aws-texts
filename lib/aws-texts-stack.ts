import * as cdk from '@aws-cdk/core';
import { PolicyDocument, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { RestApi, Integration, IntegrationType, IntegrationOptions } from '@aws-cdk/aws-apigateway';
import { EventBus, IEventBus, Rule } from '@aws-cdk/aws-events';
import { Function, Runtime, Code, IFunction } from '@aws-cdk/aws-lambda';
import { LambdaFunction, SfnStateMachine } from '@aws-cdk/aws-events-targets';
import { StateMachine, Wait, WaitTime } from '@aws-cdk/aws-stepfunctions';
import { LambdaInvoke } from '@aws-cdk/aws-stepfunctions-tasks';

export class AwsTextsStack extends cdk.Stack {
  lambda: IFunction;
  eventBus: IEventBus;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.lambda = new Function(this, 'Lambda', {
      runtime: Runtime.NODEJS_14_X,
      code: Code.fromAsset('lib/lambda-fns'),
      handler: 'oi.handler'
    });

    const waitTask = new Wait(this, 'WaitUntil', {
      time: WaitTime.duration(cdk.Duration.seconds(10))//.timestampPath('$.detail.at')
    });
    const lambdaTask = new LambdaInvoke(this, 'LambdaInvoke', {
      lambdaFunction: this.lambda,
      outputPath: '$.Payload'
    });

    const stateMachine = new StateMachine(this, 'TextStateMachine', {
      definition: waitTask.next(lambdaTask)
    });

    this.eventBus = new EventBus(this, 'TextEventBus', { 
      eventBusName: 'TextEventBus' 
    });
    new Rule(this, 'LambdaProcessorRule', {
      eventBus: this.eventBus,
      eventPattern: {source:['apigateway'], detailType:['text']},
      targets: [
      //  new LambdaFunction(this.lambda)
        new SfnStateMachine(stateMachine)
      ]
    });
    const apigwEventBusRole = new Role(this, 'ApiGatewayEventBusRole', {
      assumedBy: new ServicePrincipal('apigateway'),
      inlinePolicies: {
        putEvents: new PolicyDocument({
          statements: [new PolicyStatement({
            actions: ['events:PutEvents'],
            resources: [this.eventBus.eventBusArn]
          })]
        })
      }
    });

    const options:IntegrationOptions = {
      credentialsRole: apigwEventBusRole,
      requestParameters: {
        "integration.request.header.X-Amz-Target": "'AWSEvents.PutEvents'",
        "integration.request.header.Content-Type": "'application/x-amz-json-1.1'"
      },
      requestTemplates: {
        "application/json": `{"Entries": [
          {"Source": "apigateway", 
          "Detail": "$util.escapeJavaScript($input.body)", 
          "DetailType": "text", 
          "EventBusName": "${this.eventBus.eventBusName}"}]}`
      },
      integrationResponses: [{
        statusCode: "200",
        responseTemplates: {"application/json": ""}
      },
      {
        statusCode: "400",
        responseTemplates: {"application/json": JSON.stringify(
          { state: 'error', message: "$util.escapeJavaScript($input.path('$.errorMessage'))" })}
      }]
    };
    const gw = new RestApi(this, 'TextApiGateway', {deployOptions: {stageName: 'dev'}});
    gw.root.addMethod('POST', new Integration({
        type: IntegrationType.AWS,
        uri: 'arn:aws:apigateway:us-east-1:events:path//',
        integrationHttpMethod: 'POST',
        options:options
      }),
      {
        methodResponses: [{statusCode: "200"}, {statusCode: "400"}]
      }
    );

    // The code that defines your stack goes here

    // example resource
    // const queue = new sqs.Queue(this, 'AwsTextsQueue', {
    //   visibilityTimeout: cdk.Duration.seconds(300)
    // });
  }
}
