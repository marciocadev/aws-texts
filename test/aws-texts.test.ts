import { Template } from '@aws-cdk/assertions';
import { App, Stack } from '@aws-cdk/core';
import { Function, CfnFunction } from '@aws-cdk/aws-lambda';
import { AwsTextsStack } from '../lib/aws-texts-stack';
import { CfnEventBus } from '@aws-cdk/aws-events';

// example test. To run these tests, uncomment this file along with the
// example resource in lib/aws-texts-stack.ts

test('Snapshot', () => {
    // GIVEN
    const app = new App();
    // WHEN
    const stack = new AwsTextsStack(app, 'MyTestStack');
    // THEN
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
});

describe('Stack tests', () => {
    let app:App;
    let stack:AwsTextsStack;
    let template:Template;

    beforeAll(() => {
        // GIVEN
        app = new App();
        // WHEN
        stack = new AwsTextsStack(app, 'MyTestStack');
        // THEN
        template = Template.fromStack(stack);
    });

    test('EventBus Created', () => {
        template.findResources('AWS::Events::EventBus');
        template.hasResourceProperties('AWS::Events::EventBus', {
            Name: 'TextEventBus'
        })
    });

    test('Event Rule Created', () => {
        const lambdaLogicalId = stack.resolve((stack.dominantLanguage.node.defaultChild as CfnFunction).logicalId)
        console.log(lambdaLogicalId);

        const eventBusLogicalId = stack.resolve((stack.eventBus.node.defaultChild as CfnEventBus).logicalId);

        template.resourceCountIs('AWS::Events::Rule', 1);
        template.findResources('AWS::Events::Rule');
        template.hasResourceProperties('AWS::Events::Rule', {
            State: 'ENABLED',
            EventPattern: {
                source: ['apigateway'],
                'detail-type': ['text']
            },
            EventBusName: {
                Ref: eventBusLogicalId
            }
            // Targets: [{
            //     Arn: {
            //         'Fn::GetAtt': [
            //             'LambdaD247545B',
            //             'Arn'
            //         ]
            //     },
            //     Id: 'Target0'
            // }]
        });
    });

    test('RespApi Created', () => {
        template.findResources('AWS::ApiGateway::RestApi');
        template.hasResourceProperties('AWS::ApiGateway::RestApi', {
            Name: 'TextApiGateway'
        });
    });
});
