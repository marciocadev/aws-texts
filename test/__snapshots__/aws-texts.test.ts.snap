// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "TextApiGatewayEndpoint1B713D84": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "TextApiGateway437DBC5F",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "TextApiGatewayDeploymentStagedev47E27826",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901bArtifactHash07DBFB6B": Object {
      "Description": "Artifact hash for asset \\"43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901b\\"",
      "Type": "String",
    },
    "AssetParameters43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901bS3Bucket9439CB58": Object {
      "Description": "S3 bucket for asset \\"43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901b\\"",
      "Type": "String",
    },
    "AssetParameters43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901bS3VersionKey9B6B2FEB": Object {
      "Description": "S3 key for asset version \\"43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901b\\"",
      "Type": "String",
    },
    "AssetParametersb9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8dArtifactHash36BC2B83": Object {
      "Description": "Artifact hash for asset \\"b9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8d\\"",
      "Type": "String",
    },
    "AssetParametersb9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8dS3Bucket090BF488": Object {
      "Description": "S3 bucket for asset \\"b9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8d\\"",
      "Type": "String",
    },
    "AssetParametersb9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8dS3VersionKeyC90DC4E2": Object {
      "Description": "S3 key for asset version \\"b9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8d\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AWSSDKV3FE048A04": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "nodejs14.x",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901bS3Bucket9439CB58",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901bS3VersionKey9B6B2FEB",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters43f29a69bc57dc930645800fa35ae7a0a945e740b0020c53f6f65b4687e8901bS3VersionKey9B6B2FEB",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ApiGatewayEventBusRole78CD632E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "events:PutEvents",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "TextEventBus50F77630",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "putEvents",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DominantLanguageA68A3AD3": Object {
      "DependsOn": Array [
        "DominantLanguageServiceRoleDefaultPolicy7B2F070E",
        "DominantLanguageServiceRole9E3D3018",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersb9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8dS3Bucket090BF488",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8dS3VersionKeyC90DC4E2",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb9bae1bfe487ad371468707a625ee7797ac5f60cf114f17e89d83c1aede0ce8dS3VersionKeyC90DC4E2",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "AWSSDKV3FE048A04",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "DominantLanguageServiceRole9E3D3018",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "DominantLanguageServiceRole9E3D3018": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DominantLanguageServiceRoleDefaultPolicy7B2F070E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "comprehend:DetectDominantLanguage",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DominantLanguageServiceRoleDefaultPolicy7B2F070E",
        "Roles": Array [
          Object {
            "Ref": "DominantLanguageServiceRole9E3D3018",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaProcessorRuleC28EF8EB": Object {
      "Properties": Object {
        "EventBusName": Object {
          "Ref": "TextEventBus50F77630",
        },
        "EventPattern": Object {
          "detail-type": Array [
            "text",
          ],
          "source": Array [
            "apigateway",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "TextStateMachine95BF0DB2",
            },
            "Id": "Target0",
            "RoleArn": Object {
              "Fn::GetAtt": Array [
                "TextStateMachineEventsRole7E52DA46",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TextApiGateway437DBC5F": Object {
      "Properties": Object {
        "Name": "TextApiGateway",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "TextApiGatewayAccount54FA72E1": Object {
      "DependsOn": Array [
        "TextApiGateway437DBC5F",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "TextApiGatewayCloudWatchRoleE5030C8C",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "TextApiGatewayCloudWatchRoleE5030C8C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TextApiGatewayDeployment754CD196c3c1adbc4ea08c0597ce4b42e4ae4328": Object {
      "DependsOn": Array [
        "TextApiGatewayPOST0201D577",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "TextApiGateway437DBC5F",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "TextApiGatewayDeploymentStagedev47E27826": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "TextApiGatewayDeployment754CD196c3c1adbc4ea08c0597ce4b42e4ae4328",
        },
        "RestApiId": Object {
          "Ref": "TextApiGateway437DBC5F",
        },
        "StageName": "dev",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "TextApiGatewayPOST0201D577": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "ApiGatewayEventBusRole78CD632E",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "ResponseTemplates": Object {
                "application/json": "",
              },
              "StatusCode": "200",
            },
            Object {
              "ResponseTemplates": Object {
                "application/json": "{\\"state\\":\\"error\\",\\"message\\":\\"$util.escapeJavaScript($input.path('$.errorMessage'))\\"}",
              },
              "StatusCode": "400",
            },
          ],
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'application/x-amz-json-1.1'",
            "integration.request.header.X-Amz-Target": "'AWSEvents.PutEvents'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{\\"Entries\\": [
          {\\"Source\\": \\"apigateway\\", 
          \\"Detail\\": \\"$util.escapeJavaScript($input.body)\\", 
          \\"DetailType\\": \\"text\\", 
          \\"EventBusName\\": \\"",
                  Object {
                    "Ref": "TextEventBus50F77630",
                  },
                  "\\"}]}",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": "arn:aws:apigateway:us-east-1:events:path//",
        },
        "MethodResponses": Array [
          Object {
            "StatusCode": "200",
          },
          Object {
            "StatusCode": "400",
          },
        ],
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "TextApiGateway437DBC5F",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "TextApiGateway437DBC5F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TextEventBus50F77630": Object {
      "Properties": Object {
        "Name": "TextEventBus",
      },
      "Type": "AWS::Events::EventBus",
    },
    "TextStateMachine95BF0DB2": Object {
      "DependsOn": Array [
        "TextStateMachineRoleDefaultPolicy7A727CB5",
        "TextStateMachineRoleA1896FF6",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"WaitUntil\\",\\"States\\":{\\"WaitUntil\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":10,\\"Next\\":\\"LambdaInvoke\\"},\\"LambdaInvoke\\":{\\"End\\":true,\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"OutputPath\\":\\"$.Payload\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "DominantLanguageA68A3AD3",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "TextStateMachineRoleA1896FF6",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "TextStateMachineEventsRole7E52DA46": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TextStateMachineEventsRoleDefaultPolicyC69D252C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "TextStateMachine95BF0DB2",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TextStateMachineEventsRoleDefaultPolicyC69D252C",
        "Roles": Array [
          Object {
            "Ref": "TextStateMachineEventsRole7E52DA46",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TextStateMachineRoleA1896FF6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TextStateMachineRoleDefaultPolicy7A727CB5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "DominantLanguageA68A3AD3",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TextStateMachineRoleDefaultPolicy7A727CB5",
        "Roles": Array [
          Object {
            "Ref": "TextStateMachineRoleA1896FF6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
